dm/pkg/shardddl/optimism/lock.go

```diff
@@ -871,3 +714,419 @@ func GetColumnName(lockID, ddl string, tp ast.AlterTableType) (string, error) {
+	l.conflictTables = make(map[string]map[string]map[string]schemacmp.Table)
+	l.finalTables = make(map[string]map[string]map[string]schemacmp.Table)
+	for source, schemaTables := range l.tables {
+		l.finalTables[source] = make(map[string]map[string]schemacmp.Table)
+		for schema, tables := range schemaTables {
+			l.finalTables[source][schema] = make(map[string]schemacmp.Table)
+			for table, ti := range tables {
+				l.finalTables[source][schema][table] = ti
+			}
+		}
+	}
+	// update for current conflict DDL
+	l.addConflictTable(source, schema, table, postTable)
+	l.finalTables[source][schema][table] = postTable
+
+	return l.noConflictForFinalTables()
+}
+
+// judge whether all conflict tables has no conflict.
+func (l *Lock) noConflictForConflictTables() bool {
+	if _, err := l.joinConflictTables(); err != nil {
+		return false
+	}
+	if !l.allConflictTableSmaller() {
+		return false
+	}
+	if !l.allConflictTableLarger() {
+		return false
+	}
+	return true
+}
+
+// judge whether all final tables has no conflict.
+func (l *Lock) noConflictForFinalTables() bool {
+	if _, err := l.joinFinalTables(); err != nil {
+		return false
+	}
+	if !l.allFinalTableSmaller() {
+		return false
+	}
+	if !l.allFinalTableLarger() {
+		return false
+	}
+	return true
+}
+
+func (l *Lock) addConflictTable(source, schema, table string, ti schemacmp.Table) {
+	if _, ok := l.conflictTables[source]; !ok {
+		l.conflictTables[source] = make(map[string]map[string]schemacmp.Table)
+	}
+	if _, ok := l.conflictTables[source][schema]; !ok {
+		l.conflictTables[source][schema] = make(map[string]schemacmp.Table)
+	}
+	if _, ok := l.conflictTables[source][schema]; !ok {
+		l.conflictTables[source][schema] = make(map[string]schemacmp.Table)
+	}
+	l.conflictTables[source][schema][table] = ti
+}
+
+func (l *Lock) removeConflictTable(source, schema, table string) {
+	if _, ok := l.conflictTables[source]; !ok {
+		return
+	}
+	if _, ok := l.conflictTables[source][schema]; !ok {
+		return
+	}
+	delete(l.conflictTables[source][schema], table)
+	if len(l.conflictTables[source][schema]) == 0 {
+		delete(l.conflictTables[source], schema)
+	}
+	if len(l.conflictTables[source]) == 0 {
+		delete(l.conflictTables, source)
+	}
+}
+
+// resolveTables reset conflictTables and copy tables from final tables.
+func (l *Lock) resolveTables() {
+	l.conflictTables = make(map[string]map[string]map[string]schemacmp.Table)
+	for source, schemaTables := range l.finalTables {
+		for schema, tables := range schemaTables {
+			for table, ti := range tables {
+				l.tables[source][schema][table] = ti
+			}
+		}
+	}
+}
+
+// redirectForConflictTables put redirect Ops for all conflict tables.
+func (l *Lock) redirectForConflictTables() error {
+	for source, schemaTables := range l.conflictTables {
+		for schema, tables := range schemaTables {
+			for table := range tables {
+				op := NewOperation(l.ID, l.Task, source, schema, table, nil, ConflictResolved, "", false, nil)
```

Comment:
i'm a bit concerned about the performance issue of etcd here, it's a triple neste loop 😭, can we batch these ops?

Reply:
later pr with #3829